File: sdk/typescript/src/process.ts
//! Process supervision — spawn, feed, stream, and stop the `sentry` daemon.
//!
//! `Sentry` maps a typed `SentryConfig` to the daemon's `A3S_SENTRY_*` env vars, spawns it,
//! writes events to its stdin, and exposes the audit NDJSON on stdout as a typed async iterator.
//!
//! IMPORTANT: the daemon emits an audit line ONLY for blocks, escalations, and flagged-but-allowed
//! events. Plain benign allows are counted, not printed — so `decisions()` yields exactly the
//! audited (non-allow) decisions, never one per submitted event.

import { type ChildProcessByStdio, spawn } from "node:child_process";
import type { Readable, Writable } from "node:stream";
import * as readline from "node:readline";
import { type Audit, parseAudit } from "./types.js";
import type { SentryEvent } from "./events.js";

/** stdin piped (write events), stdout piped (read audit), stderr inherited (daemon logs). */
type SentryChild = ChildProcessByStdio<Writable, Readable, null>;

/** Typed daemon configuration. Each field maps to one `A3S_SENTRY_*` env var. */
export interface SentryConfig {
  /** The binary path/name. Default `"sentry"`. */
  bin?: string;
  /** Extra L1 rules (HCL) — `A3S_SENTRY_POLICY`. Built-ins always apply; hot-reloaded (~2s). */
  policy?: string;
  /** Observer egress deny-file — `A3S_SENTRY_EGRESS_DENY`. */
  egressDeny?: string;
  /** Observer file deny-file — `A3S_SENTRY_FILE_DENY`. */
  fileDeny?: string;
  /** Observer exec deny-file — `A3S_SENTRY_EXEC_DENY`. */
  execDeny?: string;
  /** L2 OpenAI-compatible endpoint base — `A3S_SENTRY_LLM_URL`. Enables L2. */
  llmUrl?: string;
  /** L2 model name — `A3S_SENTRY_LLM_MODEL`. */
  llmModel?: string;
  /** L2 bearer token — `A3S_SENTRY_LLM_KEY`. */
  llmKey?: string;
  /** L3 a3s-code binary — `A3S_SENTRY_AGENT_BIN`. Enables L3. */
  agentBin?: string;
  /** L3 security-skills directory — `A3S_SENTRY_SKILLS`. */
  skills?: string;
  /** Unresolved escalations BLOCK — `A3S_SENTRY_FAIL_CLOSED=1`. */
  failClosed?: boolean;
  /** Speculate threshold severity — `A3S_SENTRY_SPECULATE` (e.g. `"high"`). */
  speculate?: string;
  /** L2 request timeout in seconds — `A3S_SENTRY_LLM_TIMEOUT`. */
  llmTimeout?: number;
  /** L3 investigation timeout in seconds — `A3S_SENTRY_AGENT_TIMEOUT`. */
  agentTimeout?: number;
  /** L2/L3 worker thread count — `A3S_SENTRY_WORKERS`. */
  workers?: number;
  /** Escalation queue depth — `A3S_SENTRY_QUEUE`. */
  queue?: number;
  /** Judge + audit, but never write a deny-file — `A3S_SENTRY_DRY_RUN=1`. */
  dryRun?: boolean;
  /** Serve Prometheus `/metrics` + `/healthz` — `A3S_SENTRY_METRICS_ADDR` (e.g. `"0.0.0.0:9100"`). */
  metricsAddr?: boolean | string;
}

/** Build the `A3S_SENTRY_*` env overlay from a config. Exported for inspection/testing. */
export function configToEnv(config: SentryConfig): Record<string, string> {
  const env: Record<string, string> = {};
  const set = (key: string, value: string | undefined): void => {
    if (value !== undefined && value !== "") {
      env[key] = value;
    }
  };
  set("A3S_SENTRY_POLICY", config.policy);
  set("A3S_SENTRY_EGRESS_DENY", config.egressDeny);
  set("A3S_SENTRY_FILE_DENY", config.fileDeny);
  set("A3S_SENTRY_EXEC_DENY", config.execDeny);
  set("A3S_SENTRY_LLM_URL", config.llmUrl);
  set("A3S_SENTRY_LLM_MODEL", config.llmModel);
  set("A3S_SENTRY_LLM_KEY", config.llmKey);
  set("A3S_SENTRY_AGENT_BIN", config.agentBin);
  set("A3S_SENTRY_SKILLS", config.skills);
  set("A3S_SENTRY_SPECULATE", config.speculate);
  if (config.llmTimeout !== undefined) set("A3S_SENTRY_LLM_TIMEOUT", String(config.llmTimeout));
  if (config.agentTimeout !== undefined)
    set("A3S_SENTRY_AGENT_TIMEOUT", String(config.agentTimeout));
  if (config.workers !== undefined) set("A3S_SENTRY_WORKERS", String(config.workers));
  if (config.queue !== undefined) set("A3S_SENTRY_QUEUE", String(config.queue));
  if (config.failClosed) env["A3S_SENTRY_FAIL_CLOSED"] = "1";
  if (config.dryRun) env["A3S_SENTRY_DRY_RUN"] = "1";
  if (typeof config.metricsAddr === "string" && config.metricsAddr !== "") {
    env["A3S_SENTRY_METRICS_ADDR"] = config.metricsAddr;
  }
  return env;
}

/** Default grace period (ms) to wait for a clean drain-and-exit before SIGTERM. */
const DEFAULT_STOP_TIMEOUT_MS = 10_000;

/**
 * A supervised `sentry` daemon process.
 *
 * Lifecycle: `start()` → `submit(event)` (repeat) + consume `decisions()` → `stop()`.
 * Implements `Symbol.asyncDispose`, so `await using s = new Sentry(cfg); s.start();` stops it
 * automatically at scope exit.
 */
export class Sentry {
  private readonly config: SentryConfig;
  private child: SentryChild | undefined;
  private exited: Promise<number | null> | undefined;

  constructor(config: SentryConfig = {}) {
    this.config = config;
  }

  /** Whether the daemon is currently running. */
  get running(): boolean {
    return this.child !== undefined && this.child.exitCode === null && !this.child.killed;
  }

  /** The underlying process id, if started. */
  get pid(): number | undefined {
    return this.child?.pid;
  }

  /** Spawn the daemon, inheriting `process.env` with the config overlay applied. */
  start(): void {
    if (this.child !== undefined) {
      throw new Error("Sentry already started");
    }
    const bin = this.config.bin ?? "sentry";
    const env = { ...process.env, ...configToEnv(this.config) };
    // stdin: write events · stdout: read audit NDJSON · stderr: inherit (daemon logs to stderr).
    const child: SentryChild = spawn(bin, [], {
      env,
      stdio: ["pipe", "pipe", "inherit"],
    });
    this.child = child;
    this.exited = new Promise((resolve) => {
      child.once("exit", (code) => resolve(code));
    });
  }

  /** Write one event to the daemon's stdin (`event.toLine() + "\n"`). */
  submit(event: SentryEvent): void {
    const child = this.requireChild();
    child.stdin.write(event.toLine() + "\n");
  }

  /**
   * Stream typed audit decisions from the daemon's stdout, one per line. Blanks and unparseable
   * lines are skipped. Yields exactly the audited (non-allow) decisions — benign allows are not
   * emitted by the daemon. The iterator ends when the daemon's stdout closes (after `stop()`).
   */
  async *decisions(): AsyncIterableIterator<Audit> {
    const child = this.requireChild();
    const rl = readline.createInterface({ input: child.stdout, crlfDelay: Infinity });
    for await (const line of rl) {
      const trimmed = line.trim();
      if (trimmed === "") {
        continue;
      }
      let audit: Audit;
      try {
        audit = parseAudit(trimmed);
      } catch {
        continue; // skip stray non-JSON lines
      }
      yield audit;
    }
  }

  /**
   * Stop the daemon gracefully: close stdin so it reaches EOF, drains its queue, and exits.
   * Waits up to `timeoutMs` for a clean exit, then SIGTERMs. Resolves with the exit code.
   */
  async stop(timeoutMs: number = DEFAULT_STOP_TIMEOUT_MS): Promise<number | null> {
    const child = this.child;
    if (child === undefined || this.exited === undefined) {
      return null;
    }
    // EOF on stdin → the daemon finishes the queue and exits on its own.
    child.stdin.end();

    let timer: NodeJS.Timeout | undefined;
    const timeout = new Promise<"timeout">((resolve) => {
      timer = setTimeout(() => resolve("timeout"), timeoutMs);
      timer.unref?.();
    });
    const result = await Promise.race([this.exited.then(() => "exited" as const), timeout]);
    if (timer !== undefined) {
      clearTimeout(timer);
    }
    if (result === "timeout" && this.running) {
      child.kill("SIGTERM");
    }
    return this.exited;
  }

  /** `await using` support — stops the daemon at scope exit. */
  async [Symbol.asyncDispose](): Promise<void> {
    await this.stop();
  }

  private requireChild(): SentryChild {
    if (this.child === undefined) {
      throw new Error("Sentry not started — call start() first");
    }
    return this.child;
  }
}

File: sdk/python/a3s_sentry/process.py
"""Supervise a sentry process: feed it events, stream typed decisions back."""

from __future__ import annotations

import asyncio
import json
import os
from dataclasses import dataclass
from typing import AsyncIterator, Dict, Optional

from .events import Event
from .types import Audit


@dataclass
class SentryConfig:
    """Typed configuration → sentry's ``A3S_SENTRY_*`` environment. Unset fields are omitted."""

    bin: str = "sentry"  # the sentry binary (name on PATH or a path)
    policy: Optional[str] = None  # ACL policy file (hot-reloaded)
    egress_deny: Optional[str] = None
    file_deny: Optional[str] = None
    exec_deny: Optional[str] = None
    llm_url: Optional[str] = None
    llm_model: Optional[str] = None
    llm_key: Optional[str] = None
    agent_bin: Optional[str] = None
    skills: Optional[str] = None
    fail_closed: bool = False
    speculate: Optional[str] = None  # severity threshold, e.g. "high"
    llm_timeout: Optional[int] = None
    agent_timeout: Optional[int] = None
    workers: Optional[int] = None
    queue: Optional[int] = None
    dry_run: bool = False
    metrics_addr: Optional[str] = None  # e.g. "127.0.0.1:9100"

    def env(self) -> Dict[str, str]:
        e: Dict[str, str] = {}

        def put(key: str, value: object) -> None:
            if value is not None:
                e[key] = str(value)

        put("A3S_SENTRY_POLICY", self.policy)
        put("A3S_SENTRY_EGRESS_DENY", self.egress_deny)
        put("A3S_SENTRY_FILE_DENY", self.file_deny)
        put("A3S_SENTRY_EXEC_DENY", self.exec_deny)
        put("A3S_SENTRY_LLM_URL", self.llm_url)
        put("A3S_SENTRY_LLM_MODEL", self.llm_model)
        put("A3S_SENTRY_LLM_KEY", self.llm_key)
        put("A3S_SENTRY_AGENT_BIN", self.agent_bin)
        put("A3S_SENTRY_SKILLS", self.skills)
        if self.fail_closed:
            e["A3S_SENTRY_FAIL_CLOSED"] = "1"
        put("A3S_SENTRY_SPECULATE", self.speculate)
        put("A3S_SENTRY_LLM_TIMEOUT", self.llm_timeout)
        put("A3S_SENTRY_AGENT_TIMEOUT", self.agent_timeout)
        put("A3S_SENTRY_WORKERS", self.workers)
        put("A3S_SENTRY_QUEUE", self.queue)
        if self.dry_run:
            e["A3S_SENTRY_DRY_RUN"] = "1"
        put("A3S_SENTRY_METRICS_ADDR", self.metrics_addr)
        return e


class Sentry:
    """A running sentry process. Use it as an async context manager::

        async with Sentry(SentryConfig(egress_deny="egress.txt")) as s:
            await s.submit(Event.egress(1, "169.254.169.254", 80))
            async for audit in s.decisions():
                print(audit.decision.verdict, audit.subject)

    Sentry emits an audit line only for blocks / escalations / flagged events — plain benign allows
    are counted, not printed — so :meth:`decisions` yields exactly the noteworthy ones.
    """

    def __init__(self, config: Optional[SentryConfig] = None) -> None:
        self.config = config or SentryConfig()
        self._proc: Optional[asyncio.subprocess.Process] = None

    async def start(self) -> "Sentry":
        env = dict(os.environ)
        env.update(self.config.env())
        self._proc = await asyncio.create_subprocess_exec(
            self.config.bin,
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
            env=env,
        )
        return self

    async def submit(self, event: Event) -> None:
        proc = self._require()
        assert proc.stdin is not None
        proc.stdin.write((event.to_line() + "\n").encode())
        await proc.stdin.drain()

    async def decisions(self) -> AsyncIterator[Audit]:
        proc = self._require()
        assert proc.stdout is not None
        async for raw in proc.stdout:
            line = raw.strip()
            if not line:
                continue
            try:
                obj = json.loads(line)
            except json.JSONDecodeError:
                continue
            yield Audit.from_json(obj)

    async def stop(self, timeout: float = 10.0) -> None:
        """Graceful stop: close stdin (EOF → sentry drains the queue and exits), then wait."""
        proc = self._proc
        if proc is None:
            return
        if proc.stdin is not None and not proc.stdin.is_closing():
            proc.stdin.close()
        try:
            await asyncio.wait_for(proc.wait(), timeout=timeout)
        except asyncio.TimeoutError:
            proc.terminate()
            await proc.wait()
        finally:
            self._proc = None

    def _require(self) -> asyncio.subprocess.Process:
        if self._proc is None:
            raise RuntimeError("sentry is not running; call start() (or use 'async with')")
        return self._proc

    async def __aenter__(self) -> "Sentry":
        return await self.start()

    async def __aexit__(self, *_exc: object) -> None:
        await self.stop()
